indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentTransactionalInsertAction.java

```diff
@@ -99,14 +102,26 @@ public DataSourceMetadata getEndMetadata()
   @Override
   public SegmentPublishResult perform(Task task, TaskActionToolbox toolbox) throws IOException
   {
-    toolbox.verifyTaskLocks(task, segments);
+    TaskActionPreconditions.checkLockCoversSegments(task, toolbox.getTaskLockbox(), segments);
+
+    for (DataSegment segment : segments) {
+      if (!toolbox.getTaskLockbox().upgrade(task, segment.getInterval()).isOk()) {
```

Comment:
This may not work as you intend. I believe `lockbox.upgrade(task, interval)` only works if the lock interval is exactly `interval`, but it might not be in this case. For example, HadoopIndexTask acquires locks for a wider interval than its individual segments (it computes an umbrella interval and locks that).

Reply:
```lockbox.upgrade()``` actually works if there is only one lock whose interval contains the given interval. Do you think this makes sense?

Reply:
It's not consistent with how `unlock` works so it's a little strange in that regard. I guess it's fine, but the javadocs for the methods `lock`, `unlock`, `upgrade`, and `downgrade` should describe how this works, since otherwise it will get confusing.

Reply:
Good point, but in the latest patch, I simplified the logic around critical section and removed upgrade/downgrade methods.